src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,201 @@
+  // an empty type, _present is false, otherwise, _data is canonical
+  class CanonicalizedTypeIntPrototype {
+  public:
+    bool _present; // whether this is an empty set
+    TypeIntPrototype<S, U> _data;
+
+    bool empty() const {
+      return !_present;
+    }
+
+    static CanonicalizedTypeIntPrototype make_empty() {
+      return {false, {}};
+    }
+  };
+
+  CanonicalizedTypeIntPrototype canonicalize_constraints() const;
+  int normalize_widen(int w) const;
+#ifdef ASSERT
+  bool contains(S v) const;
+  void verify_constraints() const;
+#endif // ASSERT
+};
+
+// Various helper functions for TypeInt/TypeLong operations
+class TypeIntHelper {
+public:
+  // Calculate the cardinality of a TypeInt/TypeLong ignoring the bits
+  // constraints, the return value is the cardinality minus 1 to not overflow
+  // with the bottom type
+  template <class S, class U>
+  static U cardinality_from_bounds(const RangeInt<S>& srange, const RangeInt<U>& urange) {
+    static_assert(S(-1) < S(0), "");
+    static_assert(U(-1) > U(0), "");
+    static_assert(sizeof(S) == sizeof(U), "");
+
+    if (U(srange._lo) == urange._lo) {
+      // srange is the same as urange
+      assert(U(srange._hi) == urange._hi, "");
+      // The cardinality is (hi - lo + 1), we return the result minus 1
+      return urange._hi - urange._lo;
+    }
+
+    // srange intersects with urange in 2 intervals [srange._lo, urange._hi]
+    // and [urange._lo, srange._hi]
+    // The cardinality is (uhi - lo + 1) + (hi - ulo + 1), we return the result
+    // minus 1
+    return (urange._hi - U(srange._lo)) + (U(srange._hi) - urange._lo) + U(1);
+  }
+
+  template <class S, class U>
+  using make_type_t = const Type* (*)(const TypeIntPrototype<S, U>&, int, bool);
+
+  template <class CT, class S, class U>
+  static const Type* int_type_xmeet(const CT* i1, const Type* t2, make_type_t<S, U> make, bool dual);
+
+  template <class CT>
+  static bool int_type_is_equal(const CT* t1, const CT* t2) {
+    return t1->_lo == t2->_lo && t1->_hi == t2->_hi &&
+           t1->_ulo == t2->_ulo && t1->_uhi == t2->_uhi &&
+           t1->_bits._zeros == t2->_bits._zeros && t1->_bits._ones == t2->_bits._ones;
+  }
+
+  template <class CT>
+  static bool int_type_is_subset(const CT* super, const CT* sub) {
+    return super->_lo <= sub->_lo && super->_hi >= sub->_hi &&
+           super->_ulo <= sub->_ulo && super->_uhi >= sub->_uhi &&
+           // All bits that are known in super must also be known to be the same
+           // value in sub, &~ (and not) is the same as a set subtraction on bit
+           // sets
+           (super->_bits._zeros &~ sub->_bits._zeros) == 0 && (super->_bits._ones &~ sub->_bits._ones) == 0;
+  }
+
+  template <class CT>
+  static const Type* int_type_widen(const CT* nt, const CT* ot, const CT* lt);
+
+  template <class CT>
+  static const Type* int_type_narrow(const CT* nt, const CT* ot);
```

Comment:
```suggestion
  static const Type* int_type_widen(const CT* new_type, const CT* old_type, const CT* limit_type);

  template <class CT>
  static const Type* int_type_narrow(const CT* new_type, const CT* old_type, const CT* limit_type);
```
For consistency with its definition.